README.md
p5-anttweakbar
==============

Perl5 bindings for [AntTweakBar](http://anttweakbar.sourceforge.net)

[![Build Status](https://travis-ci.org/PerlGameDev/AntTweakBar.svg?branch=master)](https://travis-ci.org/PerlGameDev/AntTweakBar)


Changes
Revision history for Perl extension AntTweakBar.

        - Improved documentation (added mention of ready to use
                examples)

0.03  Tue Jul 29 22:38:22 FET 2014
        - Allow pass definition for button/variable as hashref
	- Fix typos in documentation
	- Update links to PerlGameDev github account

0.02  Sat Jul 26 23:45:00 FET 2014
        - Improved documentation about exported constants (GH#1)

0.01  Mon Jun 23 02:17:34 FET 2014
	- original version; created by h2xs 1.23 with options
		-A -n AntTweakBar


Build.PL
use strict;
use warnings;

use Module::Build;
use Devel::CheckLib;

use Alien::AntTweakBar;
use Alien::SDL;

my $opengl_lib = $^O eq 'MSWin32' ? 'opengl32' : 'GL';
check_lib_or_exit( lib => $opengl_lib );

my $atb_inc = Alien::AntTweakBar->config('cflags');
my $atb_libs = Alien::AntTweakBar->config('libs');

my $sdl_inc = Alien::SDL->config('cflags');
my $sdl_libs = Alien::SDL->config('libs');

my @all_libs = (
    ( $^O ne 'MSWin32' ? ('-x', 'c++') : () ),
    $atb_libs, $sdl_libs, "-l$opengl_lib", '-lstdc++',
);

my $all_c_flags = join(' ', $atb_inc, $sdl_inc, '-DTW_STATIC');
my $all_libs = join(' ', @all_libs);

my $build = Module::Build->new(
    module_name        => 'AntTweakBar',
    dist_version_from  => 'lib/AntTweakBar.pm',
    dist_author        => 'DMOL, dmol@cpan.org',
    dynamic_config     => 1,
    license            => 'perl',
    configure_requires => {
        'perl'                    => '5.008',
        'Module::Build'           => '0.42',
        'Alien::AntTweakBar'      => '0.02',
        'Alien::SDL'              => '1.444',
        'Devel::CheckLib'         => '1.01',
    },
    requires           => {
      'perl'               => 5.008,
      'Carp'               => 0,
      'Alien::AntTweakBar' => '0.02',
      'Alien::SDL'         => '1.444',
    },
    test_requires => {
        'Test::Fatal'    => 0,
        'Test::More'     => 0,
        'Test::Warnings' => 0,
    },
    extra_compiler_flags => $all_c_flags,
    extra_linker_flags   => $all_libs,
    c_source             => 'src',
    needs_compiler       => 1,
    add_to_cleanup     => [ 'AntTweakBar-*', 'build_done' ],
    meta_merge => {
        resources  => {
            bugtracker => 'https://github.com/PerlGameDev/AntTweakBar/issues',
            repository => 'https://github.com/PerlGameDev/AntTweakBar',
        }
    },

);

$build->create_build_script;


Makefile.PL
use strict;
use warnings;

use ExtUtils::MakeMaker;
use Devel::CheckLib;

use Alien::AntTweakBar;
use Alien::SDL;

check_lib_or_exit( lib => 'GL' );

my $atb_inc = Alien::AntTweakBar->config('cflags');
my $atb_libs = Alien::AntTweakBar->config('libs');

my $sdl_inc = Alien::SDL->config('cflags');
my $sdl_libs = Alien::SDL->config('libs');

my $all_inc = "$atb_inc $sdl_inc";
my $all_libs = "$atb_libs $sdl_libs -lGL";

print "using libs: $all_libs\n";

WriteMakefile(
    NAME              => 'AntTweakBar',
    VERSION_FROM      => 'lib/AntTweakBar.pm',
    ABSTRACT_FROM     => 'lib/AntTweakBar.pm',
    CONFIGURE_REQUIRES => {
        'Alien::AntTweakBar'  => '0.02_06',
        'Alien::SDL'          => '1.444',
        'ExtUtils::MakeMaker' => '6.52',
        'Devel::CheckLib'     => '1.01',
    },
    PREREQ_PM         => {
        'Carp'               => 0,
        'Exporter'           => 0,
        'XSLoader'           => 0,
        'Alien::AntTweakBar' => '0.02_06',
    },
    TEST_REQUIRES     => {
        'Test::More'     => 0,
        'Test::Fatal'    => 0,
        'Test::Warnings' => 0,
    },
    META_MERGE        => {
        "meta-spec" => { version => 2 },
        resources => {
            repository => {
                type => 'git',
                url => 'git@github.com:basiliscos/p5-anttweakbar.git',
                web => 'https://github.com/basiliscos/p5-anttweakbar',
            },

        },

    },
    AUTHOR            => 'Ivan Baidakou <dmol@gmx.com>',
    LICENSE           => 'perl',
    CCFLAGS           => '-g',
    LD                => '$(CXX)', # AntTweakBar is compiled via c++
    LIBS              => $all_libs,
    DEFINE            => '', # e.g., '-DHAVE_SOMETHING'
    INC               => "-I. $all_inc",
	# Un-comment this if you add C files to link with later:
    # OBJECT            => '$(O_FILES)', # link all the C files too
);


eg/simple-glut.pl
#!/usr/bin/env perl

# This is almost one-to-one perl's rewrite of C-original
# http://anttweakbar.sourceforge.net/doc/tools:anttweakbar:examples#twsimpleglut
# http://sourceforge.net/p/anttweakbar/code/ci/master/tree/examples/TwSimpleGLUT.c

use 5.12.0;
use strict;
use warnings;

use OpenGL qw/:all/;
use AntTweakBar qw/:all/;
use AntTweakBar::Type;
use List::MoreUtils qw/pairwise/;
use List::Util qw/reduce/;
use Data::Dump qw/dump/;
use Time::HiRes qw/tv_interval gettimeofday/;

sub display;

sub reshape {
    my ($width, $height) = @_;
    glViewport(0, 0, $width, $height);
    glMatrixMode(GL_PROJECTION);
    glLoadIdentity;
    gluPerspective(40, $width/$height, 1, 10);
    glMatrixMode(GL_MODELVIEW);
    glLoadIdentity;
    gluLookAt(0,0,5, 0,0,0, 0,1,0);
    glTranslatef(0, 0.6, -1);

    AntTweakBar::window_size($width, $height);
}


glutInit;
glutInitDisplayMode(GLUT_DOUBLE | GLUT_RGB | GLUT_DEPTH);
glutInitWindowSize(640, 480);
glutCreateWindow("[perl] AntTweakBar simple example using GLUT");

AntTweakBar::init(TW_OPENGL);

glutDisplayFunc(\&display);
glutReshapeFunc(\&reshape);
glutMouseFunc(\&AntTweakBar::eventMouseButtonGLUT);
glutMotionFunc(\&AntTweakBar::eventMouseMotionGLUT);
glutPassiveMotionFunc(\&AntTweakBar::eventMouseMotionGLUT);
glutKeyboardFunc(\&AntTweakBar::eventKeyboardGLUT);
glutSpecialFunc(\&AntTweakBar::eventSpecialGLUT);
AntTweakBar::GLUTModifiersFunc(\&glutGetModifiers);

reshape(640, 750);

sub set_quaternion_from_axis_angle {
    my ($axis, $angle) = @_;
    my $sina2 = sin($angle * 0.5);
    my $norm = sqrt( reduce { $a + $b} map { $_ * $_} @$axis );
    my $q = [];
    @$q = map { $sina2 * $_ / $norm } @$axis;
    $q->[3] = cos( $angle * 0.5 );
    return $q;
}

sub convert_quaternion_t_omatrix {
    my $q = shift;
    my $m = [];
    my $yy2 = 2.0 * $q->[1] * $q->[1];
    my $xy2 = 2.0 * $q->[0] * $q->[1];
    my $xz2 = 2.0 * $q->[0] * $q->[2];
    my $yz2 = 2.0 * $q->[1] * $q->[2];
    my $zz2 = 2.0 * $q->[2] * $q->[2];
    my $wz2 = 2.0 * $q->[3] * $q->[2];
    my $wy2 = 2.0 * $q->[3] * $q->[1];
    my $wx2 = 2.0 * $q->[3] * $q->[0];
    my $xx2 = 2.0 * $q->[0] * $q->[0];

    $m->[0*4+0] = - $yy2 - $zz2 + 1.0;
    $m->[0*4+1] = $xy2 + $wz2;
    $m->[0*4+2] = $xz2 - $wy2;
    $m->[0*4+3] = 0;
    $m->[1*4+0] = $xy2 - $wz2;
    $m->[1*4+1] = - $xx2 - $zz2 + 1.0;
    $m->[1*4+2] = $yz2 + $wx2;
    $m->[1*4+3] = 0;
    $m->[2*4+0] = $xz2 + $wy2;
    $m->[2*4+1] = $yz2 - $wx2;
    $m->[2*4+2] = - $xx2 - $yy2 + 1.0;
    $m->[2*4+3] = 0;
    $m->[3*4+0] = $m->[3*4+1] = $m->[3*4+2] = 0;
    $m->[3*4+3] = 1;

    return $m;
}

sub multiply_quaternions {
    my ($q1, $q2) = @_;
    my $qr = [];
	$qr->[0] = $q1->[3]*$q2->[0] + $q1->[0]*$q2->[3] + $q1->[1]*$q2->[2] - $q1->[2]*$q2->[1];
	$qr->[1] = $q1->[3]*$q2->[1] + $q1->[1]*$q2->[3] + $q1->[2]*$q2->[0] - $q1->[0]*$q2->[2];
	$qr->[2] = $q1->[3]*$q2->[2] + $q1->[2]*$q2->[3] + $q1->[0]*$q2->[1] - $q1->[1]*$q2->[0];
	$qr->[3] = $q1->[3]*$q2->[3] - ($q1->[0]*$q2->[0] + $q1->[1]*$q2->[1] + $q1->[2]*$q2->[2]);
    return $qr;
}

my @shape_drawings = (
    Teapot => { function => \&glutSolidTeapot, arguments => [ 1.0 ] },
    Torus  => { function => \&glutSolidTorus,  arguments => [ 0.3, 1.0, 16, 32] },
    Cone   => { function => \&glutSolidCone,   arguments => [ 1.0, 1.5, 64, 4] },
);

# Create some 3D objects (stored in display lists)
for my $i ( 0 .. @shape_drawings/2-1 ) {
    my $shape_id = ($i * 2) + 1; # shape_id shouldn't be zero
    my $defintion = $shape_drawings[$i*2+1];
    my $drawer = $defintion->{function };
    my $args   = $defintion->{arguments};
    glNewList($shape_id, GL_COMPILE);
    $drawer->(@$args);
    glEndList;
}

my $shape_type = AntTweakBar::Type->new(
    "ShapeType",
    { map { ($shape_drawings[$_*2] => $_*2) } (0 .. @shape_drawings/2-1) },
);

# variables
my $angle            = 0.8;
my $axis             = [ 0.7, 0.7, 0.0 ];
my $zoom             = 1.0;
my $rotation         = set_quaternion_from_axis_angle($axis, $angle);
my $rotate_start     = set_quaternion_from_axis_angle($axis, $angle);
my $auto_rotation    = 0;
my $rotate_time      = [ gettimeofday ];
my $light_multiplier = 1.0;
my $light_direction  = [ -0.57735, -0.57735, -0.57735 ];
my $material_ambient = [ 0.5, 0.0, 0.0];
my $material_diffuse = [ 1.0, 1.0, 0.0];
my $shape            = 2; # torus

sub display {
    glClearColor(0, 0, 0, 1);
    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);

    glEnable(GL_DEPTH_TEST);
    glDisable(GL_CULL_FACE);
    glEnable(GL_NORMALIZE);

    # set light
    glEnable(GL_LIGHTING);
    glEnable(GL_LIGHT0);
    my $ambient_light = OpenGL::Array->new_list(
        GL_FLOAT, (0.4 * $light_multiplier) x 3, 1.0);
    my $diffuse_light = OpenGL::Array->new_list(
        GL_FLOAT, (0.8 * $light_multiplier) x 3, 1.0);
    glLightfv_c(GL_LIGHT0, GL_AMBIENT, $ambient_light->ptr);
    glLightfv_c(GL_LIGHT0, GL_DIFFUSE, $diffuse_light->ptr);
    my $light_position = OpenGL::Array->new_list(
        GL_FLOAT, map { $_ * -1 } @$light_direction, 0.0);
    glLightfv_c(GL_LIGHT0, GL_POSITION, $light_position->ptr);

    # set material
    my $oga_material_ambient = OpenGL::Array->new_list(
        GL_FLOAT, @$material_ambient);
    my $oga_material_diffuse = OpenGL::Array->new_list(
        GL_FLOAT, @$material_diffuse);
    glMaterialfv_c(GL_FRONT_AND_BACK, GL_AMBIENT, $oga_material_ambient->ptr);
    glMaterialfv_c(GL_FRONT_AND_BACK, GL_DIFFUSE, $oga_material_diffuse->ptr);

    # Rotate and draw shape
    glPushMatrix;
    {
        glTranslatef(0.5, -0.3, 0.0);
        if ($auto_rotation) {
            my $axis = [0.0, 1.0, 0.0];
            my $elapsed = tv_interval( $rotate_time, [gettimeofday]);
            my $angle = $elapsed;
            my $q = set_quaternion_from_axis_angle($axis, $angle);
            $rotation = multiply_quaternions($rotate_start, $q);
        }
        my $m = convert_quaternion_t_omatrix($rotation);
        glMultMatrixf_p(@$m);
        glScalef($zoom, $zoom, $zoom);
        glCallList($shape + 1); # shape_id = shape + 1
    }
    glPopMatrix;

    AntTweakBar::draw;
    glutSwapBuffers;
    glutPostRedisplay;
}

my $bar = AntTweakBar->new(
    "TweakBar & Perl",
    size  => '200 400',
    color => '96 216 224'
);
$bar->add_variable(
    mode       => 'rw',
    name       => "Zoom",
    type       => 'number',
    value      => \$zoom,
    definition => " min=0.01 max=2.5 step=0.01 keyIncr=z keyDecr=Z help='Scale the object (1=original size).' ",
);
$bar->add_variable(
    mode       => 'rw',
    name       => "ObjRotation",
    type       => 'quaternion',
    value      => \$rotation,
    definition => " label='Object rotation' opened=true help='Change the object orientation.' ",
);
$bar->add_variable(
    mode       => 'rw',
    name       => "AutoRotate",
    type       => 'bool',
    cb_read    => sub { $auto_rotation },
    cb_write   => sub {
        $auto_rotation = shift;
        if ($auto_rotation) {
            $rotate_time = [ gettimeofday ];
            @$rotate_start = @$rotation;
            $bar->set_variable_params('ObjRotation', readonly => 'true');
        }
        $bar->set_variable_params('ObjRotation', readonly => 'false');
    },
    definition => " label='Auto-rotate' key=space help='Toggle auto-rotate mode.' ",
);
$bar->add_variable(
    mode       => 'rw',
    name       => "Multiplier",
    type       => 'number',
    value      => \$light_multiplier,
    definition => " label='Light booster' min=0.1 max=4 step=0.02 keyIncr='+' keyDecr='-' help='Increase/decrease the light power.' ",
);
$bar->add_variable(
    mode       => 'rw',
    name       => "LightDir",
    type       => 'direction',
    value      => \$light_direction,
    definition => " label='Light direction' opened=true help='Change the light direction.' ",
);
$bar->add_variable(
    mode       => 'rw',
    name       => "Ambient",
    type       => 'color3f',
    value      => \$material_ambient,
    definition => " group='Material' ",
);
$bar->add_variable(
    mode       => 'rw',
    name       => "Diffuse",
    type       => 'color3f',
    value      => \$material_diffuse,
    definition => " group='Material' ",
);
$bar->add_variable(
    mode       => 'rw',
    name       => "Shape",
    type       => $shape_type,
    value      => \$shape,
    definition => " keyIncr='<' keyDecr='>' help='Change object shape.' ",
);

glutMainLoop;


lib/AntTweakBar/Type.pm
package AntTweakBar::Type;

use 5.12.0;

use Carp;
use Alien::AntTweakBar;
use AntTweakBar;


=head1 NAME

AntTweakBar::Type - User-defined variable types (enumerations) for AntTweakBar

=head1 SYNOPSIS

  my $framework_type = AntTweakBar::Type->new(
    "framework_type",
    [qw/Mojolicious Kelp Dancer Catalyst/],
  );

  my $fw_index = 1; # default will be Kelp
  $bar->add_variable(
    mode       => 'rw',
    name       => "used_framework",
    type       => $framework_type,
    value      => \$fw_index,
  );

  my $gender_type = AntTweakBar::Type->new(
    "gender_type",
    { male => 1, female => 2},
  );


=head1 DESCRIPTION

C<AntTweakBar::Type> allows to insert into bar variables of custom type.
Currently only enumerations are supported.

An perl variable must be refernce to integer; so, the type definition
must be eithe array or hashref with integer values.

=cut

sub new {
    my ($class, $name, $value) = @_;
    my $hash = ref($value) eq 'ARRAY'
        ? { map { ($_ => $value->[$_]) } (0 .. @$value-1) }
        : ref($value) eq 'HASH'
        ? { map { $value->{$_} => $_ } keys %$value }
        : die("New type value should be either hash or array reference");
    my $type_id = AntTweakBar::_register_enum($name, $hash);
    my $self = {
        _name    => $name,
        _type_id => $type_id,
    };
    return bless $self => $class;
}

sub name {
    shift->{_name};
}

1;


lib/AntTweakBar.pm
package AntTweakBar;

use 5.12.0;
use strict;
use warnings;

use Carp;
use Alien::AntTweakBar;

require Exporter;

our @ISA = qw(Exporter);

our @constants =
    qw/
          TW_OPENGL
          TW_OPENGL_CORE
          TW_DIRECT3D9
          TW_DIRECT3D10
          TW_DIRECT3D11
      /;

our %EXPORT_TAGS = (
    'all' => [ qw(init window_size terminate), @constants ],
    'constants' => \@constants,
);

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(

);

our $VERSION = '0.05_01';

=head1 NAME

AntTweakBar - Perl bindings for AntTweakBar

=head1 SYNOPSIS

  use AntTweakBar qw/:all/;
  use SDL::Events;

  # Setup part: link AntTweakBar with your OpenGL/SDL system

  AntTweakBar::init(TW_OPENGL);
  AntTweakBar::window_size($width, $height);

  # in your main rendering routine
  sub display {
    AntTweakBar::draw;
  }

  sub process_events {
    SDL::Events::pump_events;
    my $event = $self->sdl_event;
    while (SDL::Events::poll_event($event)) {
        ...;
        AntTweakBar::eventSDL($event);
    }
  }


  # define bars with variables
  my $bar = AntTweakBar->new(
    "TweakBar & Perl",
    size  => '200 400',
    color => '96 216 224'
  );
  my $enabled = 0;
  $bar->add_variable(
    mode       => 'rw',
    name       => "Enabled",
    type       => 'bool',
    value      => \$enabled
);


=head1 DESCRIPTION

=for HTML
<p>
  <img src="https://raw.githubusercontent.com/PerlGameDev/AntTweakBar/master/eg/available-properties.png" style="max-width:100%;">
  <img src="https://raw.githubusercontent.com/PerlGameDev/AntTweakBar/master/eg/simple-glut.gif" style="max-width:100%;">
</p>


AntTweakBar (see L<http://anttweakbar.sourceforge.net/>) is nice tiny
GUI library for OpenGL/SDL/DirectX applications.

If you are in hurry to evaluate it then intall AntTweakBar and try
examples from C<eg> directory within the distribution.

To display AntTweakBar in your OpenGL/SDL application you should do the
following:

=over

=item intialize AntTweakBar(s):

  AntTweakBar::init(TW_OPENGL);
  AntTweakBar::window_size($width, $height);

=item draw AntTweakBar(s):

  AntTweakBar::draw;

=item let AntTweakBar(s) respond to user interactions:

  AntTweakBar::eventSDL($event);

=item create AntTweakBar instance(s)

  my $bar = AntTweakBar->new("TweakBar");

=item add variables into the $bar

  my $value = 3.14;
  $bar->add_variable(
    mode       => 'rw',
    name       => "Enabled",
    type       => 'integer',
    value      => \$value
  );

=back

See working examples in the C<eg> directoctory within the distribution.

=head2 EXPORT

Constants only


=head1 CONSTANTS

The following constants let AntTweakBar know which graphic system do
you use, to know how to render itself

=head2 TW_OPENGL

Render using plain old OpenGL

=head2 TW_OPENGL_CORE

Render AntTweakBar using OpenGL core profile, which excludes deprecated
OpenGL functions. See: L<https://en.wikipedia.org/wiki/OpenGL_4#OpenGL_3.2>

=head2 TW_DIRECT3D9 (not implemented)

=head2 TW_DIRECT3D10 (not implemented)

=head2 TW_DIRECT3D11 (not implemented)

=cut

require XSLoader;
XSLoader::load('AntTweakBar', $VERSION);


=head1 METHODS

=head2 new

  my $bar = AntTweakBar->new(
    "TweakBar & Perl",
    size  => '200 400',
    color => '96 216 224'
  );

  my $another_bar = AntTweakBar->new(
    "Misc.",
  );


Creates new AntTweakBar instance. Optionally the list of strings of
bar-related parameters can be provided. See the list of available
at L<http://anttweakbar.sourceforge.net/doc/tools:anttweakbar:twbarparamsyntax>.

=cut

sub new {
    my ($class, $name, %bar_params) = @_;
    croak "AntTweakBar name should be specified"
        unless defined $name;
    my $self = {
        _name    => $name,
        _bar_ptr => _create( $name ),
    };
    bless $self => $class;
    $self->set_bar_params(%bar_params);
    return $self;
}

sub DESTROY {
    my $self = shift;
    _destroy($self->{_bar_ptr});
}


sub _as_definition_string {
    my $d = shift;
    return join(' ', map {
        my $key = $_;
        my $value = $d->{$key};
        $value =~ s/'/\\'/g;
        "$key='$value'";
    } sort keys %$d);
}

=head2 add_button

  $bar->add_button(
    name       => 'my_btn',
    cb         => sub { say "clicked!" },
    definition => "label='Click me!'",    # optional
  );

  $bar->add_button(
    name       => 'my_btn',
    cb         => sub { say "clicked!" },
    definition => { # optional
      label => 'Click me!'
    }
  );


The definition parameters are the same as for variable. See
L<http://anttweakbar.sourceforge.net/doc/tools:anttweakbar:varparamsyntax#parameters>.

=cut


sub add_button {
    my ($self, %args) = @_;

    my $name        = $args{name      };
    my $cb          = $args{cb        };
    my $definition  = $args{definition} // "";

    croak "Button name should be specified"
        unless defined $name;
    croak "Button callback should be specified"
        if(!defined($cb) || ref($cb) ne 'CODE');
    $definition = _as_definition_string($definition)
        if ($definition && ref($definition) eq 'HASH');

    _add_button($self->{_bar_ptr}, $name, $cb, $definition);
}


=head2 add_separator

  $bar->add_separator('separator-name');

=cut

sub add_separator {
    my ($self, $name, $definition) = @_;
    croak "Separator name should be specified"
        unless defined $name;

    $definition //= "";
    $definition = _as_definition_string($definition)
        if ($definition && ref($definition) eq 'HASH');

    _add_separator($self->{_bar_ptr}, $name, $definition);
}


=head2 add_variable

  my $zoom = 1.0;
  $bar->add_variable(
    mode       => 'rw',
    name       => "Zoom",
    type       => 'number',
    value      => \$zoom,
    definition => " min=0.01 max=2.5 step=0.01 help='Bla-bla-bla.' ",
  );

  # the same, but with more perlish style in definition
  $bar->add_variable(
    mode       => 'rw',
    name       => "Zoom",
    type       => 'number',
    value      => \$zoom,
    definition => {
        min     => "0.01",
        max     => "2.5",
        step    => "0.01",
        keyIncr => 'z',
        keyDecr => 'Z',
        help    => 'Scale the object (1=original size).'
    },
  );

  my $bool = undef;
  $bar->add_variable(
    mode       => 'rw',
    name       => "bool_rw_cb",
    type       => 'bool',
    cb_read    => sub { $bool; },
    cb_write   => sub {
        $bool = shift;
        say "writing value $bool";
    }
  );

C<mode>, C<name>, C<type> are mandatory. Either C<value> or C<cb_read>
should be specified. The C<definition> and C<cb_write> are optional.

=head3 mode

The B<mode> can be C<rw> (read/write) or C<ro> (read only). The mode
specified whether the variable value could be modified via AntTweakBar.

=head3 name

Defines the unique variable name at tweakbar. Unless C<label> is
specified via C<defintion>, then C<name> also defines the visual label.
for the variable.

=head3 type

Defines the type of variable. Original types L<http://anttweakbar.sourceforge.net/doc/tools:anttweakbar:twtype>
were reduced to:

=over

=item bool

=item integer

=item number (float)

=item string

=item color3f

variable must be reference to array, consisted of 3 float values: rgb.

=item color4f

variable must be reference to array, consisted of 4 float values: rgba.

=item direction

3D-vector (direction). The variable must be reference to array, consisted of 3 float values.

=item quaternion

4D-vector (3D-object rotation). The variable must be reference to array, consisted of 4 float values.

=item custom type of L<Anttweakbar::Type>

=back

=head3 value

The B<reference> to the variable value. For complex types (e.g. quaternion) it must
also be an B<reference> to array of 3 numbers.

=head3 cb_read

Closure, that returns the actual value of variable.

=head3 cb_write($value)

Closure, that is been invoked when user sets new value to the variable.
If C<cb_write> is undefined, then the variable considered B<readonly>.

=head3 definition

An string or hashref of values that allows additional tuning of
variable in Anttweakbar.
See L<http://anttweakbar.sourceforge.net/doc/tools:anttweakbar:varparamsyntax#parameters>
for possible values.

=cut

sub add_variable {
    my ($self, %args) = @_;

    for (qw/mode name type/) {
        croak "'$_' is mandatory argument for add_variable"
            unless exists $args{$_};
    }

    my $mode       = $args{mode      };
    my $name       = $args{name      };
    my $type       = $args{type      };
    my $value      = $args{value     };
    my $cb_read    = $args{cb_read   };
    my $cb_write   = $args{cb_write  };
    my $definition = $args{definition} // "";

    croak "Either value or callbacks should be specified"
        if ($value && ($cb_read || $cb_write));
    croak "cb_read is mandatory when value isn't specied"
        if (!$value && !$cb_read);
    croak "value should be a reference"
        if ($value && !ref($value));
    $type = $type->name if(ref($type) eq 'AntTweakBar::Type');
    $definition = _as_definition_string($definition)
        if ($definition && ref($definition) eq 'HASH');

    _add_variable($self->{_bar_ptr}, $mode, $name, $type, $value,
                  $cb_read, $cb_write, $definition);
}

=head2 remove_variable($name)

  $bar->remove_variable('Zoom');

=cut

sub remove_variable {
    my ($self, $name) = @_;
    _remove_variable($self->{_bar_ptr}, $name);
}

=head2 refresh

  $bar->refresh;

Tells Anttweakbar that variable values are possibly changed and
should be updated.

=cut

sub refresh {
    my $self = shift;
    _refresh($self->{_bar_ptr});
}

=head2 set_bar_params(%values)

  $bar->set_bar_params(
    size        => '350 700,
    valueswidth => '200'
    visible     => 'false',
  );

Updates bar definition at runtime. See
L<http://anttweakbar.sourceforge.net/doc/tools:anttweakbar:twbarparamsyntax>.

=cut

sub set_bar_params {
    my ($self, %params) = @_;
    while (my ($k, $v) = each(%params)) {
        _set_bar_parameter($self->{_bar_ptr}, $k, $v);
    }
}

=head2 set_variable_params($var_name, $var_definition)

  $bar->set_variable_params('ObjRotation', readonly => 'true');

Updates variable definition at runtime. See
L<http://anttweakbar.sourceforge.net/doc/tools:anttweakbar:varparamsyntax#parameters>

=cut

sub set_variable_params {
    my ($self, $variable, %params) = @_;
    while (my ($k, $v) = each(%params)) {
        _set_variable_parameter($self->{_bar_ptr}, $variable, $k, $v);
    }
}


=head1 INITIALIZATION AND DRAW FUNCTIONS


=head2 init

  AntTweakBar::init(TW_OPENGL);

Initializes AntTweakBar

=head2 terminate

  AntTweakBar::terminate

Uninitializes AntTweakBar

=head2 window_size

  AntTweakBar::window_size(640, 480);

Tell AntTweakBar the actual size of your window

=head2 draw

  AntTweakBar::draw;

Draw AntTweakBar just before the frame buffer is presented (swapped).



=head1 OPENGL EVENT FUNCTIONS


=head2 eventMouseButtonGLUT

  glutMouseFunc(\&AntTweakBar::eventMouseButtonGLUT);

Let AntTweakBar handles mouse button clicks

=head2 eventMouseMotionGLUT

  glutMotionFunc(\&AntTweakBar::eventMouseMotionGLUT);
  glutPassiveMotionFunc(\&AntTweakBar::eventMouseMotionGLUT);

Let AntTweakBar handles mouse movements with pressed button(s)
and passive mouse movements

=head2 eventKeyboardGLUT

  glutKeyboardFunc(\&AntTweakBar::eventKeyboardGLUT);

Let AntTweakBar handles key presses

=head2 eventSpecialGLUT

  glutSpecialFunc(\&AntTweakBar::eventSpecialGLUT);

=head2 GLUTModifiersFunc

  AntTweakBar::GLUTModifiersFunc(\&glutGetModifiers);

=head1 SDL EVENT FUNCTION

=head2 eventSDL

If you use SDL than it is more simple to let AntTweakBar process
all input-related events via single call:

  AntTweakBar::eventSDL($sdl_event);


=head1 SEE ALSO


L<Alien::AntTweakBar>, L<SDL>, L<OpenGL>, L<http://anttweakbar.sourceforge.net/>


=head1 AUTHOR

Ivan Baidakou E<lt>dmol@(gmx.com)E<gt>

=head1 CREDITS

=over 2

David Horner

=back

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014 by Ivan Baidakou

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.20.0 or,
at your option, any later version of Perl 5 you may have available.

=cut

1;
